private/bufpkg/bufmodule/bufmoduleapi/uploader.go

```diff
@@ -111,51 +111,109 @@ func (a *uploader) Upload(
 		return nil, err
 	}
 
+	var modules []*modulev1.Module
 	if uploadOptions.CreateIfNotExist() {
-		if err := a.createContentModulesIfNotExist(
+		// We must attempt to create each module one at a time, since CreateModules is atomic.
+		// For example, if contentModules contains 3 modules, a, b, and c, where a and b both
+		// already exist, calling CreateModules on all 3 at once will result in an `AlreadyExists`
+		// and `c` will not be created.
+		for _, contentModule := range contentModules {
+			modulesResponse, err := a.createContentModuleIfNotExist(
+				ctx,
+				primaryRegistry,
+				contentModule,
+				uploadOptions.CreateModuleVisibility(),
+			)
+			if err != nil {
+				return nil, err
+			}
+			modules = append(modules, modulesResponse...)
+		}
+	} else {
+		modules, err = a.validateContentModulesExist(
 			ctx,
 			primaryRegistry,
 			contentModules,
-			uploadOptions.CreateModuleVisibility(),
-		); err != nil {
+		)
+		if err != nil {
 			return nil, err
 		}
-	} else {
-		if err := a.validateContentModulesExist(
+	}
+
+	var v1beta1ProtoUploadRequestContents []*modulev1beta1.UploadRequest_Content
+	if len(uploadOptions.Labels()) > 0 {
+		// While the API allows different labels per reference, we don't expose this through
+		// the use of the `--label` flag, so all references will have the same labels.
+		// We just pre-compute them now.
+		v1beta1ProtoScopedLabelRefs := slicesext.Map(
+			uploadOptions.Labels(),
+			labelNameToV1Beta1ProtoScopedLabelRef,
+		)
+		// Maintains ordering, important for when we create bufmodule.Commit objects below.
+		v1beta1ProtoUploadRequestContents, err = slicesext.MapError(
+			contentModules,
+			func(module bufmodule.Module) (*modulev1beta1.UploadRequest_Content, error) {
+				return getV1Beta1ProtoUploadRequestContent(
+					ctx,
+					v1beta1ProtoScopedLabelRefs,
+					primaryRegistry,
+					module,
+				)
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if uploadOptions.BranchOrDraft() != "" {
+		if len(contentModules) > 1 {
+			return nil, fmt.Errorf("--branch and --draft are disallowed for use when pushing a workspace with more than one module")
+		}
+		// We know that there is only one module we are uploading contents for.
+		v1beta1ProtoUploadRequestContent, err := getV1Beta1ProtoUploadRequestContent(
 			ctx,
+			[]*modulev1beta1.ScopedLabelRef{
+				labelNameToV1Beta1ProtoScopedLabelRef(uploadOptions.BranchOrDraft()),
+			},
 			primaryRegistry,
-			contentModules,
-		); err != nil {
+			contentModules[0],
```

Comment:
Are we validating somewhere that `len(contentModules) > 0`?

Reply:
+1

Reply:
I changed the error condition above to `if len(contentModules) != 1`, since it doesn't really make sense otherwise. There is a case to be made that `Upload` should have an early return `if len(contentModules) == 0 { return nil, nil }` since there isn't anything to upload. This is also a pretty unlikely case/you're more likely to run into `no .proto files` error while building the workspace if that is happening.